### PR TITLE
update mallocMC

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Update `mallocMC`. The interface changes in `mallocMC` make it necessary  that `PMacc` and `PIConGPU` also need some interface changes. All in all the change is that we need to pass the mallocMC heap to the constructor of the particle classes.

- update third party mallocMC
- PIConGPU
  - delete `mallocMCconfig.hpp`
  - add param file `mallocMC.param`
  - `main.cu` - remove mallocMC include
  - ADIOS: change `MallocMCBuffer` to `MallocMCBuffer<DeviceHeap>`
  - `MySimulation.hpp`
    - add member `deviceHeap` (shared pointer)
    - change `MallocMCBuffer` to `MallocMCBuffer<DeviceHeap>`
  - `Particles<>` - extend constructor parameters
  - `ParticlesFunctors` - consider `Particles<>` interface changes
- PMacc
  - `ParticlesBase`:
    - add template parameter `T_DeviceHeap`
    - add device heap to constructor parameter
  - `MallocMCBuffer` - add template parameter `T_DeviceHeap`
  - `ParticlesBox` - add member for mallocMC device heap

Tests:
- [x] KHI heating electrons/positrons

### CODE REVIEW

File: src/libPMacc/include/particles/memory/buffers/ParticlesBuffer.hpp [hunk 1]
```diff
@@ -169,12 +174,13 @@ class ParticlesBuffer
...
      * Constructor.
      *
+     * @param deviceHeap device heap memory allocator
      * @param layout number of cell per dimension
      * @param superCellSize size of one super cell
      * @param gpuMemory how many memory on device is used for this instance (in byte)
      */
-    ParticlesBuffer(DataSpace<DIM> layout, DataSpace<DIM> superCellSize) :
-    superCellSize(superCellSize), gridSize(layout), framesExchanges(nullptr)
+    ParticlesBuffer(std::shared_ptr<DeviceHeap> deviceHeap, DataSpace<DIM> layout, DataSpace<DIM> superCellSize) :
```
Comment (psychocoderHPC) [inline src/libPMacc/include/particles/memory/buffers/ParticlesBuffer.hpp]: `deviceHeap` should be a const reference

File: src/picongpu/include/particles/Particles.hpp [hunk 2]
```diff
@@ -71,14 +73,13 @@ class Particles : public ParticlesBase<
...
-    typedef ParticlesBase<SpeciesParticleDescription, MappingDesc> ParticlesBaseType;
-    typedef typename ParticlesBaseType::BufferType BufferType;
+    typedef ParticlesBase<SpeciesParticleDescription, MappingDesc, DeviceHeap> ParticlesBaseType;
     typedef typename ParticlesBaseType::FrameType FrameType;
     typedef typename ParticlesBaseType::FrameTypeBorder FrameTypeBorder;
     typedef typename ParticlesBaseType::ParticlesBoxType ParticlesBoxType;
 
 
-    Particles(MappingDesc cellDescription, SimulationDataId datasetID);
+    Particles(std::shared_ptr<DeviceHeap> heap, MappingDesc cellDescription, SimulationDataId datasetID);
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/Particles.hpp]: `heap` should be a reference

File: src/picongpu/include/particles/ParticlesFunctors.hpp [hunk 3]
```diff
@@ -75,10 +77,18 @@ struct CreateSpecies
     typedef T_SpeciesName SpeciesName;
     typedef typename SpeciesName::type SpeciesType;
 
-    template<typename T_StorageTuple, typename T_CellDescription>
-    HINLINE void operator()(T_StorageTuple& tuple, T_CellDescription* cellDesc) const
+    template<typename T_StorageTuple, typename T_DeviceHeap, typename T_CellDescription>
+    HINLINE void operator()(
+        T_StorageTuple& tuple,
+        std::shared_ptr<T_DeviceHeap> deviceHeap,
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ParticlesFunctors.hpp]: should be a const ref

File: src/picongpu/include/particles/ParticlesFunctors.hpp [hunk 4]
```diff
@@ -88,14 +98,17 @@ struct CallCreateParticleBuffer
     typedef T_SpeciesName SpeciesName;
     typedef typename SpeciesName::type SpeciesType;
 
-    template<typename T_StorageTuple>
-    HINLINE void operator()(T_StorageTuple& tuple) const
+    template<typename T_StorageTuple, typename T_DeviceHeap>
+    HINLINE void operator()(
+        T_StorageTuple& tuple,
+        std::shared_ptr<T_DeviceHeap> deviceHeap
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ParticlesFunctors.hpp]: const ref

File: src/picongpu/include/simulationControl/MySimulation.hpp [hunk 5]
```diff
@@ -335,8 +336,12 @@ class MySimulation : public SimulationHelper<simDim>
             this->bremsstrahlungPhotonAngle.init();
         }
 
+        /* Create an empty allocator. This one is resized after all exchanges
+         * for particles are created */
+        deviceHeap.reset(new DeviceHeap(0));
+
         ForEach<VectorAllSpecies, particles::CreateSpecies<bmpl::_1>, MakeIdentifier<bmpl::_1> > createSpeciesMemory;
-        createSpeciesMemory(forward(particleStorage), cellDescription);
+        createSpeciesMemory(forward(particleStorage), deviceHeap, cellDescription);
```
Comment (psychocoderHPC) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: forward deviceHeap

File: src/picongpu/include/simulationControl/MySimulation.hpp [hunk 6]
```diff
@@ -362,11 +367,11 @@ class MySimulation : public SimulationHelper<simDim>
...
 
         // initializing the heap for particles
-        mallocMC::initHeap(heapSize);
-        this->mallocMCBuffer = new MallocMCBuffer();
+        deviceHeap->destructiveResize(heapSize);
+        this->mallocMCBuffer = new MallocMCBuffer<DeviceHeap>(deviceHeap);
 
         ForEach<VectorAllSpecies, particles::CallCreateParticleBuffer<bmpl::_1>, MakeIdentifier<bmpl::_1> > createParticleBuffer;
-        createParticleBuffer(forward(particleStorage));
+        createParticleBuffer(forward(particleStorage), deviceHeap);
```
Comment (psychocoderHPC) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: forward deviceHeap

**Comment by psychocoderHPC** [comment]
ready for a review

**Comment by ax3l** [comment]
thanks for grouping the commits so nicely, looks good to me!

**Pull request merged**